cmd/contour/serve.go

```diff
@@ -352,6 +383,99 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	}
 	g.Add(debugsvc.Start)
 
+	// Leader Election starts here.
+
+	leaderOK := make(chan struct{})
+	deposed := make(chan struct{})
+	leadingDisabled := make(chan struct{})
+
+	g.Add(func(stop <-chan struct{}) error {
+		log := log.WithField("context", "leaderelect")
+		if ctx.DisableLeaderElection {
+			log.Info("Leader election disabled")
+			// if leader election is disabled, signal the gRPC goroutine
+			// to start serving and block this goroutine forever.
+			close(leaderOK)
+			<-leadingDisabled
+		}
+
+		log.WithFields(logrus.Fields{
+			"configmapname":      ctx.LeaderElectionConfig.Name,
+			"configmapnamespace": ctx.LeaderElectionConfig.Namespace,
+		}).Info("Started leader election")
+
+		// Generate the event recorder to send election events to the logs.
+		// Set up the event bits
+		eventBroadcaster := record.NewBroadcaster()
+		eventBroadcaster.StartLogging(log.Infof)
+		// Broadcast election events to the config map
+		eventBroadcaster.StartRecordingToSink(&clientcorev1.EventSinkImpl{Interface: clientcorev1.New(client.CoreV1().RESTClient()).Events("")})
+		eventsScheme := runtime.NewScheme()
+		// Need an eventsScheme
+		err := coreV1.AddToScheme(eventsScheme)
+		check(err)
+		// The recorder is what will record events about the resource lock
+		recorder := eventBroadcaster.NewRecorder(eventsScheme, coreV1.EventSource{Component: "contour"})
+
+		// Figure out the resource lock ID
+		var resourceLockID string
+		resourceLockID, isPodNameEnvSet := os.LookupEnv("POD_NAME")
+		if !isPodNameEnvSet {
+			resourceLockID = uuid.New().String()
+		}
+
+		rl, err := resourcelock.New(
+			resourcelock.ConfigMapsResourceLock,
+			ctx.LeaderElectionConfig.Namespace,
+			ctx.LeaderElectionConfig.Name,
+			client.CoreV1(),
+			coordinationClient,
+			resourcelock.ResourceLockConfig{
+				Identity:      resourceLockID,
+				EventRecorder: recorder,
+			},
+		)
+		check(err)
+
+		electionCtx := context.Background()
+		le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+			Lock:          rl,
+			LeaseDuration: ctx.LeaderElectionConfig.LeaseDuration,
+			RenewDeadline: ctx.LeaderElectionConfig.RenewDeadline,
+			RetryPeriod:   ctx.LeaderElectionConfig.RetryPeriod,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(ctx context.Context) {
+					log.WithFields(logrus.Fields{
+						"lock":     rl.Describe(),
+						"identity": rl.Identity(),
+					}).Info("elected leader")
+					close(leaderOK)
+				},
+				OnStoppedLeading: func() {
+					// The context being canceled will trigger a handler that will
+					// deal with being deposed.
+					close(deposed)
+					electionCtx.Done()
```

Comment:
Unlike sync.WaitGroup, Done does not signal that the context is done, it returns a channel which will receive a signal when the context is canceled.